hal/src/rtl872x/pwm_hal.cpp

```diff
@@ -99,17 +91,19 @@ bool isPwmPin(uint16_t pin) {
 }
 
 void calculateArrAndPrescaler(uint32_t frequency, uint32_t* arr, uint8_t* prescaler) {
-    // Figure out the maximum ARR that meets the frequency requirment
-    // The higher ARR, the higher resolution we can set
+    // The PWM output frequency is calculated by the following formula
+    // Frequency = (Xtal/prescaler) / arr
     *arr = RTL_XTAL_CLOCK_HZ / frequency;
-    if (*arr > ((1 << RTL_PWM_TIM_MAX_RESOLUTION) - 1)) {
-        *arr = (1 << RTL_PWM_TIM_MAX_RESOLUTION) - 1;
-    }
-    float temp = (float)RTL_XTAL_CLOCK_HZ / frequency / *arr;
-    *prescaler = (uint8_t)(temp + 0.5);
-    if (*prescaler != 0) {
-        *prescaler -= 1;
+    uint8_t calculated_prescaler = 0;
+
+    // If the calculated ARR value is larger than will fit in the 16 bit ARR register
+    // Increase the prescaler and recalculate until we get a value that fits
+    while (*arr > ((1 << RTL_PWM_TIM_MAX_RESOLUTION) - 1) && (calculated_prescaler < MAX_PRESCALER-1)) {

```

Comment:
I think it was because of a bug found with lower resolutions.  As a compromise can we pre-calculate things, and then drop into the while() loop to make a quick tweak, instead of starting from `calculated_prescaler = 0` every time?  I'm sure this is pretty fast already, and also I guess this does get called on every call to analogWrite* so the time it takes could be important to some.

Reply:
When the calculated `arr` value was larger than 16 bits, it was truncated at 16 bits. 
Then the truncated value was used to calculate the prescaler value. This would lead to incorrect PWM frequencies, especially at low frequencies. 

For example, the default PWM frequency is 500hz. that means `arr` is calculated to be:
`40mhz / 500 hz = 80000`
which is greater than `1<<16 -1` (ie 0xFFFF) and is truncated to `0xFFFF`

When using the truncated value to calculate prescaler:
`40mhz / 500 hz / 0xFFFF = 1.22`
`1.22 + 0.5 = 1.72`
Casting to `uint8_t` rounds to `1`. Subtracting 1 from that means the prescaler should be 0.

`arr` set to `0xFFFF` and prescaler `0` gives a frequency calculation of:
`40mhz / 0xFFFF = 610 hz`
Which is not the desired 500hz output. 

This method is somewhat slower but should be more accurate.

Reply:
That makes perfect sense. 👍

Reply:
Can we maybe cache the frequency value and only re-calculate if the frequency for the pin changes?

Reply:
Determining the prescaler is only done when changing the frequency. Do we think this will be done frequently enough to need to improve performance here? 

Modifying duty cycle via `hal_pwm_update_duty_cycle_ext()` only updates the CCR register, which should be fast.

I would prefer to not make this more complicated unless needed

Reply:
It looked to me that `calculateArrAndPrescaler` was called from every call to `hal_pwm_write_with_frequency_ext` which also gets called by `hal_pwm_write*`.